primitive/src/main/java/io/atomix/primitive/impl/AbstractAsyncPrimitiveProxy.java

```diff
@@ -0,0 +1,239 @@
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected <R> CompletableFuture<R> applyOn(PartitionId partitionId, Function<S, R> operation) {
+    return getServiceProxy(partitionId).apply(operation);
+  }
+
+  /**
+   * Submits an empty operation to the owning partition for the given key.
+   *
+   * @param key       the key for which to submit the operation
+   * @param operation the operation
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected CompletableFuture<Void> acceptBy(String key, Consumer<S> operation) {
+    return getServiceProxy(key).accept(operation);
+  }
+
+  /**
+   * Submits an empty operation to the owning partition for the given key.
+   *
+   * @param key       the key for which to submit the operation
+   * @param operation the operation
+   * @param <R>       the operation result type
+   * @return A completable future to be completed with the operation result. The future is guaranteed to be completed after all
+   * {@link PrimitiveOperation} submission futures that preceded it.
+   * @throws NullPointerException if {@code operation} is null
+   */
+  protected <R> CompletableFuture<R> applyBy(String key, Function<S, R> operation) {
+    return getServiceProxy(key).apply(operation);
+  }
+
+  /**
+   * Service proxy container.
+   */
+  private class ServiceProxy<S> {
+    private final S proxy;
+    private final ServiceProxyHandler handler;
+
+    ServiceProxy(S proxy, ServiceProxyHandler handler) {
+      this.proxy = proxy;
+      this.handler = handler;
+    }
+
+    /**
+     * Invokes a void method on the underlying proxy.
+     *
+     * @param operation the operation to perform on the proxy
+     * @return the resulting void future
+     */
+    CompletableFuture<Void> accept(Consumer<S> operation) {
+      operation.accept(proxy);
+      return handler.getResultFuture();
+    }
+
+    /**
+     * Invokes a function on the underlying proxy.
+     *
+     * @param operation the operation to perform on the proxy
+     * @param <T>       the operation return type
+     * @return the future result
+     */
+    <T> CompletableFuture<T> apply(Function<S, T> operation) {
+      operation.apply(proxy);
+      return handler.getResultFuture();
+    }
+  }
+
+  /**
+   * Service proxy invocation handler.
+   * <p>
+   * The invocation handler
+   */
+  private class ServiceProxyHandler implements InvocationHandler {
+    private final PartitionProxy proxy;
+    private final ThreadLocal<CompletableFuture> future = new ThreadLocal<>();
```

Comment:
this screams trouble to me, are we super sure the future is completed / mutated always on the right thread?

Reply:
The calling thread is getting the future. It's just a trick to get a future from a proxy object. Yes, the future will always be read from the `ThreadLocal` immediately after the proxy is used.

Reply:
The alternative would be using ByteBuddy to generate a class with a different return type (`CompletableFuture`) so a synchronous method call on the proxy object can return a future, but that's a lot of overhead for something that can be done this way. The `ThreadLocal` is just to ensure multiple threads can use the proxy at the same time.

Reply:
This can also be done another way - by creating thread local proxies rather than storing thread local futures, but it's probably a distinction without much of a difference.